lib/src/git_backend.rs

```diff
@@ -356,6 +359,120 @@ impl GitBackend {
+        if contents.parents.is_empty() {
+            return Err(BackendError::Other(
+                "Cannot write a commit with no parents".into(),
+            ));
+        }
+        let mut parents = vec![];
+        for parent_id in &contents.parents {
+            if *parent_id == self.root_commit_id {
+                // Git doesn't have a root commit, so if the parent is the root commit, we don't
+                // add it to the list of parents to write in the Git commit. We also check that
+                // there are no other parents since Git cannot represent a merge between a root
+                // commit and another commit.
+                if contents.parents.len() > 1 {
+                    return Err(BackendError::Other(
+                        "The Git backend does not support creating merge commits with the root \
+                         commit as one of the parents."
+                            .into(),
+                    ));
+                }
+            } else {
+                parents.push(validate_git_object_id(parent_id)?);
+            }
+        }
+        let parents: SmallVec<[gix::ObjectId; 1]> = parents.into();
+        let extras = serialize_extras(&contents);
+
+        // If two writers write commits of the same id with different metadata, they
+        // will both succeed and the metadata entries will be "merged" later. Since
+        // metadata entry is keyed by the commit id, one of the entries would be lost.
+        // To prevent such race condition locally, we extend the scope covered by the
+        // table lock. This is still racy if multiple machines are involved and the
+        // repository is rsync-ed.
+        let (table, table_lock) = self.read_extra_metadata_table_locked()?;
+        let (git_id, commit) = loop {
+            let mut commit = gix::objs::Commit {
+                message: message.to_owned().into(),
+                tree: git_tree_id,
+                author: author.into(),
+                committer: committer.into(),
+                encoding: None,
+                parents: parents.clone(),
+                extra_headers: Default::default(),
+            };
+
+            if let Some(signing_fn) = &mut signing_fn {
+                // we don't use gix pool, but at least use their heuristic
+                let mut data = Vec::with_capacity(512);
+                commit.write_to(&mut data).unwrap();
+
+                let sig = signing_fn(&data)?;
+                commit
+                    .extra_headers
+                    .push(("gpgsig".into(), sig.clone().into()));
+                contents.secure_sig = Some(SecureSig { data, sig });
+            }
+
+            let git_id =
+                locked_repo
+                    .write_object(&commit)
+                    .map_err(|err| BackendError::WriteObject {
+                        object_type: "commit",
+                        source: Box::new(err),
+                    })?;
+
```

Comment:
Could you reorganize this patch to two or three parts? For example, 1. inline gix `commit_as()`, 2. implement function for commit signing.

It's uneasy to review code move + functionality changes.

Reply:
Sure
Although now with signing_fn being part of main `write_commit` there's no move :)